{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1071d5c7",
   "metadata": {},
   "source": [
    "# Consulta de la base de datos KEGG con Python/Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e040ec0",
   "metadata": {},
   "source": [
    "## Pregunta\n",
    "\n",
    "> ¿Cómo podemos acceder a las bases de datos de KEGG a través del módulo **Bio.KEGG** de Biopython y extraer información relevante sobre genes y vías metabólicas para su análisis bioinformático?\n",
    "\n",
    "## Objetivos\n",
    "\n",
    "### General\n",
    "> Aprender a usar el módulo **Bio.KEGG** de Biopython para consultar información biológica en la base de datos KEGG y comprender los tipos de datos que se pueden obtener.\n",
    "\n",
    "### Específicos\n",
    "1. Conocer qué es **Bio.KEGG** y para qué se utiliza.\n",
    "2. Realizar consultas básicas a KEGG para obtener información sobre genes, rutas metabólicas, compuestos, entre otros.\n",
    "3. Interpretar de manera sencilla los resultados obtenidos de las consultas.\n",
    "4. Extraer y organizar información útil para su análisis posterior o visualización."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40d8f6d5",
   "metadata": {},
   "source": [
    "## 1. ¿Qué es Bio.KEGG?\n",
    "\n",
    "De manera general, [**Bio.KEGG**](https://biopython.org/docs/1.84/Tutorial/chapter_kegg.html) es un módulo de [**Biopython**](https://biopython.org/) que permite acceder y manipular datos de la base de datos [**KEGG**](https://www.genome.jp/kegg/) directamente desde Python.  \n",
    "\n",
    "En términos simples:\n",
    "\n",
    "**KEGG** es una base de datos que nos permite entender cómo funcionan las células, los organismos y los ecosistemas a partir de información molecular. Contiene datos sobre genes, rutas metabólicas, compuestos y organismos, entre otros.\n",
    "\n",
    "**Bio.KEGG** actúa como un puente entre Python y KEGG, facilitando la consulta, descarga y análisis de estos datos.\n",
    "\n",
    "Se basa en la REST API (una forma de “pedir y recibir datos” entre programas a través de Internet, usando URLs y peticiones HTTP) de KEGG, por lo que puedes hacer consultas programáticas y obtener resultados en formatos fáciles de manejar con Python.\n",
    "\n",
    "Permite tanto consultas simples (por ejemplo, obtener la información de un gen) como operaciones más complejas (como analizar rutas metabólicas completas o relacionar genes con compuestos).\n",
    " \n",
    "En la siguiente tabla se muestran las funciones para hacer consultas a **KEGG** por medio de **Bio.KEGG**:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccc7721c",
   "metadata": {},
   "source": [
    "| Función         | Descripción                                                                 |\n",
    "|-----------------|-----------------------------------------------------------------------------|\n",
    "| `kegg_list()`   | Lista todas las bases de datos de KEGG.                                     |\n",
    "| `kegg_info()`   | Obtiene información general sobre una base de datos de KEGG.                |\n",
    "| `kegg_conv()`   | Convierte identificadores de KEGG a los de otras bases de datos.            |\n",
    "| `kegg_find()`   | Busca entradas en KEGG que coincidan con un dato de consulta.               |\n",
    "| `kegg_get()`    | Recupera información detallada de una entrada específica en KEGG.           |\n",
    "| `kegg_link()`   | Encuentra entradas en KEGG usando referencias cruzadas de otras bases.      |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e73cc4ee",
   "metadata": {},
   "source": [
    "En términos sencillos, la forma general de usar estas funciones es siempre la misma: elegir la función, indicar lo que se va a buscar y guardar el resultado en una variable.\n",
    "\n",
    "```python\n",
    "myvar = REST.<función>(<consulta>, <arg1>, <arg2>, ...)\n",
    "```\n",
    "\n",
    "- `<función>`: el nombre de la función que quieres usar (por ejemplo, kegg_list, kegg_get).\n",
    "- `<consulta>`: el dato que quieres buscar en KEGG (por ejemplo, el nombre de un organismo, un gen o una ruta).\n",
    "- `<arg1>`, `<arg2>`, ...: algunos parámetros extra que ciertas funciones necesitan.\n",
    "\n",
    "En las siguientes secciones veremos cómo usar cada una de estas funciones."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a18d95e7",
   "metadata": {},
   "source": [
    "## 2. ```kegg_list```\n",
    "\n",
    "De manera general, la función ```kegg_list``` devuelve una lista de identificadores de KEGG junto con su descripción asociada, ya sea de toda una base de datos o de un conjunto específico de elementos. Para consultar la descripción y parámetros de la función, podemos solicitar la yuda. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "66af2fcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function kegg_list in module Bio.KEGG.REST:\n",
      "\n",
      "kegg_list(database, org=None)\n",
      "    KEGG list - Entry list for database, or specified database entries.\n",
      "\n",
      "    db - database or organism (string)\n",
      "    org - optional organism (string), see below.\n",
      "\n",
      "    For the pathway and module databases the optional organism can be\n",
      "    used to restrict the results.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Importando la biblioteca REST\n",
    "from Bio.KEGG import REST\n",
    "\n",
    "# Solicitando la ayuda de kegg_list\n",
    "help(REST.kegg_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56a7bea0",
   "metadata": {},
   "source": [
    "### 2.1 ¿Qué bases de datos puedo consultar?\n",
    "\n",
    "En particular, si a ```kegg_list``` se le proporciona el parámetro **\"database\"**, se obtiene la lista de todas las bases de datos disponibles en KEGG para su consulta. Como se muestra en el siguiente código."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "008dbfec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pathway</td>\n",
       "      <td>KEGG Pathway Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>module</td>\n",
       "      <td>KEGG Module Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>network</td>\n",
       "      <td>KEGG Network Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>variant</td>\n",
       "      <td>KEGG Variant Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>disease</td>\n",
       "      <td>KEGG Disease Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "      <td>KEGG Brite Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>orthology</td>\n",
       "      <td>KEGG Orthology Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>genes</td>\n",
       "      <td>KEGG Genes Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "      <td>KEGG Genome Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>mgenes</td>\n",
       "      <td>KEGG Metagenome Genes Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>mgenome</td>\n",
       "      <td>KEGG Metagenome Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ligand</td>\n",
       "      <td>KEGG Ligand Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>compound</td>\n",
       "      <td>KEGG Compound Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>dgroup</td>\n",
       "      <td>KEGG Drug Group Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>drug</td>\n",
       "      <td>KEGG Drug Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>glycan</td>\n",
       "      <td>KEGG Glycan Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>reaction</td>\n",
       "      <td>KEGG Reaction Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>rclass</td>\n",
       "      <td>KEGG Reaction Class Database</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>enzyme</td>\n",
       "      <td>KEGG Enzyme Database</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                               1\n",
       "0     pathway           KEGG Pathway Database\n",
       "1      module            KEGG Module Database\n",
       "2     network           KEGG Network Database\n",
       "3     variant           KEGG Variant Database\n",
       "4     disease           KEGG Disease Database\n",
       "5       brite             KEGG Brite Database\n",
       "6   orthology         KEGG Orthology Database\n",
       "7       genes             KEGG Genes Database\n",
       "8      genome            KEGG Genome Database\n",
       "9      mgenes  KEGG Metagenome Genes Database\n",
       "10    mgenome        KEGG Metagenome Database\n",
       "11     ligand            KEGG Ligand Database\n",
       "12   compound          KEGG Compound Database\n",
       "13     dgroup        KEGG Drug Group Database\n",
       "14       drug              KEGG Drug Database\n",
       "15     glycan            KEGG Glycan Database\n",
       "16   reaction          KEGG Reaction Database\n",
       "17     rclass    KEGG Reaction Class Database\n",
       "18     enzyme            KEGG Enzyme Database"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Importando la biblioteca de pandas para trabajar con tablas (dataframes) y\n",
    "# y la biblioteca io, para trabajar con datos (archivos, cadenas, binarios) \"como si fueran archivos\"\n",
    "import pandas as pd\n",
    "import io\n",
    "\n",
    "# Obtener lista de bases de datos de KEGG\n",
    "result = REST.kegg_list(\"database\").read()\n",
    "\n",
    "df = pd.read_table(io.StringIO(result), header = None)\n",
    "\n",
    "display(df)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67cde5dd",
   "metadata": {},
   "source": [
    "### 2.2 Lista de todos los organismos registrados en Kegg\n",
    "\n",
    "Para consultar la lista de organismos disponibles en KEGG, utilizamos el parámetro **\"organism\"**, en la función `kegg_list`. Por ejemplo:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ac19f7b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T01001</td>\n",
       "      <td>hsa</td>\n",
       "      <td>Homo sapiens (human)</td>\n",
       "      <td>Eukaryotes;Animals;Mammals;Primates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T01005</td>\n",
       "      <td>ptr</td>\n",
       "      <td>Pan troglodytes (chimpanzee)</td>\n",
       "      <td>Eukaryotes;Animals;Mammals;Primates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T02283</td>\n",
       "      <td>pps</td>\n",
       "      <td>Pan paniscus (bonobo)</td>\n",
       "      <td>Eukaryotes;Animals;Mammals;Primates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T02442</td>\n",
       "      <td>ggo</td>\n",
       "      <td>Gorilla gorilla gorilla (western lowland gorilla)</td>\n",
       "      <td>Eukaryotes;Animals;Mammals;Primates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T01416</td>\n",
       "      <td>pon</td>\n",
       "      <td>Pongo abelii (Sumatran orangutan)</td>\n",
       "      <td>Eukaryotes;Animals;Mammals;Primates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11301</th>\n",
       "      <td>T09593</td>\n",
       "      <td>lob</td>\n",
       "      <td>Candidatus Lokiarchaeum ossiferum</td>\n",
       "      <td>Prokaryotes;Archaea;Promethearchaeota;Lokiarch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11302</th>\n",
       "      <td>T06555</td>\n",
       "      <td>psyt</td>\n",
       "      <td>Promethearchaeum syntrophicum</td>\n",
       "      <td>Prokaryotes;Archaea;Promethearchaeota;Promethe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11303</th>\n",
       "      <td>T09928</td>\n",
       "      <td>oyw</td>\n",
       "      <td>Candidatus Odinarchaeum yellowstonii</td>\n",
       "      <td>Prokaryotes;Archaea;Odinarchaeota;Odinarchaeum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11304</th>\n",
       "      <td>T03835</td>\n",
       "      <td>agw</td>\n",
       "      <td>Archaeon GW2011_AR10</td>\n",
       "      <td>Prokaryotes;Archaea;unclassified Archaea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11305</th>\n",
       "      <td>T03843</td>\n",
       "      <td>arg</td>\n",
       "      <td>Archaeon GW2011_AR20</td>\n",
       "      <td>Prokaryotes;Archaea;unclassified Archaea</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11306 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0     1                                                  2  \\\n",
       "0      T01001   hsa                               Homo sapiens (human)   \n",
       "1      T01005   ptr                       Pan troglodytes (chimpanzee)   \n",
       "2      T02283   pps                              Pan paniscus (bonobo)   \n",
       "3      T02442   ggo  Gorilla gorilla gorilla (western lowland gorilla)   \n",
       "4      T01416   pon                  Pongo abelii (Sumatran orangutan)   \n",
       "...       ...   ...                                                ...   \n",
       "11301  T09593   lob                  Candidatus Lokiarchaeum ossiferum   \n",
       "11302  T06555  psyt                      Promethearchaeum syntrophicum   \n",
       "11303  T09928   oyw               Candidatus Odinarchaeum yellowstonii   \n",
       "11304  T03835   agw                               Archaeon GW2011_AR10   \n",
       "11305  T03843   arg                               Archaeon GW2011_AR20   \n",
       "\n",
       "                                                       3  \n",
       "0                    Eukaryotes;Animals;Mammals;Primates  \n",
       "1                    Eukaryotes;Animals;Mammals;Primates  \n",
       "2                    Eukaryotes;Animals;Mammals;Primates  \n",
       "3                    Eukaryotes;Animals;Mammals;Primates  \n",
       "4                    Eukaryotes;Animals;Mammals;Primates  \n",
       "...                                                  ...  \n",
       "11301  Prokaryotes;Archaea;Promethearchaeota;Lokiarch...  \n",
       "11302  Prokaryotes;Archaea;Promethearchaeota;Promethe...  \n",
       "11303     Prokaryotes;Archaea;Odinarchaeota;Odinarchaeum  \n",
       "11304           Prokaryotes;Archaea;unclassified Archaea  \n",
       "11305           Prokaryotes;Archaea;unclassified Archaea  \n",
       "\n",
       "[11306 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Consultando los organismos disponibles en Kegg\n",
    "# Saber si el organismos x tiene base de datos en KEGG\n",
    "\n",
    "result = REST.kegg_list(database=\"organism\").read()\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None, sep=\"\\t\")\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c34515fd",
   "metadata": {},
   "source": [
    "Una vez que se obtiene la tabla, se puede filtrar la información para un organismo en particular. Por ejemplo, podemos buscar los datos relacionados con ***Drosophila melanogaster***."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ea1d06e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>492</th>\n",
       "      <td>dme</td>\n",
       "      <td>Drosophila melanogaster (fruit fly)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>der</td>\n",
       "      <td>Drosophila erecta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>494</th>\n",
       "      <td>dse</td>\n",
       "      <td>Drosophila sechellia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>495</th>\n",
       "      <td>dsi</td>\n",
       "      <td>Drosophila simulans</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>496</th>\n",
       "      <td>dya</td>\n",
       "      <td>Drosophila yakuba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>497</th>\n",
       "      <td>dan</td>\n",
       "      <td>Drosophila ananassae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>dsr</td>\n",
       "      <td>Drosophila serrata</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>dpo</td>\n",
       "      <td>Drosophila pseudoobscura</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>dpe</td>\n",
       "      <td>Drosophila persimilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>501</th>\n",
       "      <td>dmn</td>\n",
       "      <td>Drosophila miranda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>502</th>\n",
       "      <td>dwi</td>\n",
       "      <td>Drosophila willistoni</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>503</th>\n",
       "      <td>dgr</td>\n",
       "      <td>Drosophila grimshawi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>dmo</td>\n",
       "      <td>Drosophila mojavensis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>505</th>\n",
       "      <td>daz</td>\n",
       "      <td>Drosophila arizonae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>dnv</td>\n",
       "      <td>Drosophila navojoa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>507</th>\n",
       "      <td>dhe</td>\n",
       "      <td>Drosophila hydei</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>508</th>\n",
       "      <td>dvi</td>\n",
       "      <td>Drosophila virilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4130</th>\n",
       "      <td>wol</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila melanogaster</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4132</th>\n",
       "      <td>wen</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila simulans wHa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4133</th>\n",
       "      <td>wed</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila simulans wNo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4141</th>\n",
       "      <td>wdm</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila mauritiana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4144</th>\n",
       "      <td>wea</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila ananassae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4146</th>\n",
       "      <td>wey</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila yakuba</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4150</th>\n",
       "      <td>wer</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila santomea</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4153</th>\n",
       "      <td>weh</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila innubila</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4162</th>\n",
       "      <td>wev</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila pseudotak...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4164</th>\n",
       "      <td>wedc</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila aff. chau...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4171</th>\n",
       "      <td>wenw</td>\n",
       "      <td>Wolbachia endosymbiont of Drosophila willistoni</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         1                                                  2\n",
       "492    dme                Drosophila melanogaster (fruit fly)\n",
       "493    der                                  Drosophila erecta\n",
       "494    dse                               Drosophila sechellia\n",
       "495    dsi                                Drosophila simulans\n",
       "496    dya                                  Drosophila yakuba\n",
       "497    dan                               Drosophila ananassae\n",
       "498    dsr                                 Drosophila serrata\n",
       "499    dpo                           Drosophila pseudoobscura\n",
       "500    dpe                              Drosophila persimilis\n",
       "501    dmn                                 Drosophila miranda\n",
       "502    dwi                              Drosophila willistoni\n",
       "503    dgr                               Drosophila grimshawi\n",
       "504    dmo                              Drosophila mojavensis\n",
       "505    daz                                Drosophila arizonae\n",
       "506    dnv                                 Drosophila navojoa\n",
       "507    dhe                                   Drosophila hydei\n",
       "508    dvi                                 Drosophila virilis\n",
       "4130   wol  Wolbachia endosymbiont of Drosophila melanogaster\n",
       "4132   wen  Wolbachia endosymbiont of Drosophila simulans wHa\n",
       "4133   wed  Wolbachia endosymbiont of Drosophila simulans wNo\n",
       "4141   wdm    Wolbachia endosymbiont of Drosophila mauritiana\n",
       "4144   wea     Wolbachia endosymbiont of Drosophila ananassae\n",
       "4146   wey        Wolbachia endosymbiont of Drosophila yakuba\n",
       "4150   wer      Wolbachia endosymbiont of Drosophila santomea\n",
       "4153   weh      Wolbachia endosymbiont of Drosophila innubila\n",
       "4162   wev  Wolbachia endosymbiont of Drosophila pseudotak...\n",
       "4164  wedc  Wolbachia endosymbiont of Drosophila aff. chau...\n",
       "4171  wenw    Wolbachia endosymbiont of Drosophila willistoni"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Filtrando la tabla para ver solo la información de drosophila melanogaster\n",
    "# Como identificar nuestro organimos x y su letra asociada:\n",
    "\n",
    "# Filtrando la tabla para ver solo la información de drosophila melanogaster\n",
    "filtro = df[df[2].str.contains(\"drosophila\", case=False)]\n",
    "display(filtro[[1,2]])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2976283d",
   "metadata": {},
   "source": [
    "### 2.3 Lista de todos los genes de un organismo\n",
    "\n",
    "Si como parámetro de ```kegg_list``` se proporciona el **código de un organismo específico**, se obtiene la información de los **genes asociados a ese organismo**. Por ejemplo, busquemos los genes asociados a ***Drosophila melanogaster***."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f9fa1e37",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dme:Dmel_CG17636</td>\n",
       "      <td>CDS</td>\n",
       "      <td>X:complement(124370..126714)</td>\n",
       "      <td>CG17636; uncharacterized protein, isoform A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme:Dmel_CG40494</td>\n",
       "      <td>CDS</td>\n",
       "      <td>X:127449..140340</td>\n",
       "      <td>RhoGAP1A; Rho GTPase activating protein at 1A,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dme:Dmel_CR43552</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>X:133385..133502</td>\n",
       "      <td>mir-4915-RA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dme:Dmel_CG17131</td>\n",
       "      <td>CDS</td>\n",
       "      <td>X:140318..200663</td>\n",
       "      <td>tyn; trynity, isoform A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dme:Dmel_CG17707</td>\n",
       "      <td>CDS</td>\n",
       "      <td>X:complement(142731..148426)</td>\n",
       "      <td>CG17707; uncharacterized protein, isoform B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15238</th>\n",
       "      <td>dme:Dmel_CR46485</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>2R</td>\n",
       "      <td>asRNA:CR46485; antisense RNA:CR46485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15239</th>\n",
       "      <td>dme:Dmel_CR46486</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>2L</td>\n",
       "      <td>asRNA:CR46486; antisense RNA:CR46486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15240</th>\n",
       "      <td>dme:Dmel_CR46487</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>2R</td>\n",
       "      <td>asRNA:CR46487; antisense RNA:CR46487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15241</th>\n",
       "      <td>dme:Dmel_CR46524</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>3R</td>\n",
       "      <td>snRNA:7SKb; small nuclear RNA 7SK b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15242</th>\n",
       "      <td>dme:Dmel_CR46528</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>3R</td>\n",
       "      <td>asRNA:PS4; antisense RNA:PS4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15243 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      0      1                             2  \\\n",
       "0      dme:Dmel_CG17636    CDS  X:complement(124370..126714)   \n",
       "1      dme:Dmel_CG40494    CDS              X:127449..140340   \n",
       "2      dme:Dmel_CR43552  miRNA              X:133385..133502   \n",
       "3      dme:Dmel_CG17131    CDS              X:140318..200663   \n",
       "4      dme:Dmel_CG17707    CDS  X:complement(142731..148426)   \n",
       "...                 ...    ...                           ...   \n",
       "15238  dme:Dmel_CR46485  ncRNA                            2R   \n",
       "15239  dme:Dmel_CR46486  ncRNA                            2L   \n",
       "15240  dme:Dmel_CR46487  ncRNA                            2R   \n",
       "15241  dme:Dmel_CR46524  ncRNA                            3R   \n",
       "15242  dme:Dmel_CR46528  ncRNA                            3R   \n",
       "\n",
       "                                                       3  \n",
       "0            CG17636; uncharacterized protein, isoform A  \n",
       "1      RhoGAP1A; Rho GTPase activating protein at 1A,...  \n",
       "2                                            mir-4915-RA  \n",
       "3                                tyn; trynity, isoform A  \n",
       "4            CG17707; uncharacterized protein, isoform B  \n",
       "...                                                  ...  \n",
       "15238               asRNA:CR46485; antisense RNA:CR46485  \n",
       "15239               asRNA:CR46486; antisense RNA:CR46486  \n",
       "15240               asRNA:CR46487; antisense RNA:CR46487  \n",
       "15241                snRNA:7SKb; small nuclear RNA 7SK b  \n",
       "15242                       asRNA:PS4; antisense RNA:PS4  \n",
       "\n",
       "[15243 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Consultando los genes asociados a Drosophila (dme)\n",
    "result = REST.kegg_list(database=\"dme\").read()\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None, sep=\"\\t\")\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c633bfb4",
   "metadata": {},
   "source": [
    "### 2.4 Listando los registros de las bases de datos\n",
    "\n",
    "Con `kegg_list` se puede consultar las diferentes bases de datos de **KEGG**, como `compound`, `orthology`, `brite`, `pathway`, entre otras, para listar los elementos o registros que contiene cada una de ellas.  \n",
    "\n",
    "En el caso de `pathway` (rutas metabólicas), es posible realizar búsquedas específicas para obtener aquellas rutas asociadas a un organismo en particular.\n",
    "\n",
    "A continuación, veremos algunos ejemplos de cómo utilizar esta función con distintas bases de datos.  \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d78bc671",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C00001</td>\n",
       "      <td>H2O; Water</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C00002</td>\n",
       "      <td>ATP; Adenosine 5'-triphosphate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C00003</td>\n",
       "      <td>NAD+; NAD; Nicotinamide adenine dinucleotide; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C00004</td>\n",
       "      <td>NADH; DPNH; Reduced nicotinamide adenine dinuc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C00005</td>\n",
       "      <td>NADPH; TPNH; Reduced nicotinamide adenine dinu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19527</th>\n",
       "      <td>C23092</td>\n",
       "      <td>N-Hydroxycadaverine</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19528</th>\n",
       "      <td>C23093</td>\n",
       "      <td>Desferrioxamine G1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19529</th>\n",
       "      <td>C23094</td>\n",
       "      <td>Desferrioxamine E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19530</th>\n",
       "      <td>C23095</td>\n",
       "      <td>N4-Hydroxy-1-aminopropane; N-(3-aminopropyl)hy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19531</th>\n",
       "      <td>C23096</td>\n",
       "      <td>N4-Acetyl-N4-hydroxy-1-aminopropane</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19532 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                                                  1\n",
       "0      C00001                                         H2O; Water\n",
       "1      C00002                     ATP; Adenosine 5'-triphosphate\n",
       "2      C00003  NAD+; NAD; Nicotinamide adenine dinucleotide; ...\n",
       "3      C00004  NADH; DPNH; Reduced nicotinamide adenine dinuc...\n",
       "4      C00005  NADPH; TPNH; Reduced nicotinamide adenine dinu...\n",
       "...       ...                                                ...\n",
       "19527  C23092                                N-Hydroxycadaverine\n",
       "19528  C23093                                 Desferrioxamine G1\n",
       "19529  C23094                                  Desferrioxamine E\n",
       "19530  C23095  N4-Hydroxy-1-aminopropane; N-(3-aminopropyl)hy...\n",
       "19531  C23096                N4-Acetyl-N4-hydroxy-1-aminopropane\n",
       "\n",
       "[19532 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Consultando la base de datos de compuestos\n",
    "\n",
    "result = REST.kegg_list(database=\"compound\").read()\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None, sep=\"\\t\")\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "405268d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>K00001</td>\n",
       "      <td>E1.1.1.1, adh; alcohol dehydrogenase [EC:1.1.1.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>K00002</td>\n",
       "      <td>AKR1A1, adh; alcohol dehydrogenase (NADP+) [EC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>K00003</td>\n",
       "      <td>hom; homoserine dehydrogenase [EC:1.1.1.3]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K00004</td>\n",
       "      <td>BDH, butB; (R,R)-butanediol dehydrogenase / me...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>K00005</td>\n",
       "      <td>gldA; glycerol dehydrogenase [EC:1.1.1.6]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27878</th>\n",
       "      <td>K28619</td>\n",
       "      <td>MPP6; MAGUK p55 subfamily member 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27879</th>\n",
       "      <td>K28620</td>\n",
       "      <td>MPP7; MAGUK p55 subfamily member 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27880</th>\n",
       "      <td>K28621</td>\n",
       "      <td>CABLES; CDK5 and ABL1 enzyme substrate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27881</th>\n",
       "      <td>K28622</td>\n",
       "      <td>LRRTM4; leucine-rich repeat transmembrane neur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27882</th>\n",
       "      <td>K28623</td>\n",
       "      <td>MRNIP; MRN complex-interacting protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27883 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                                                  1\n",
       "0      K00001  E1.1.1.1, adh; alcohol dehydrogenase [EC:1.1.1.1]\n",
       "1      K00002  AKR1A1, adh; alcohol dehydrogenase (NADP+) [EC...\n",
       "2      K00003         hom; homoserine dehydrogenase [EC:1.1.1.3]\n",
       "3      K00004  BDH, butB; (R,R)-butanediol dehydrogenase / me...\n",
       "4      K00005          gldA; glycerol dehydrogenase [EC:1.1.1.6]\n",
       "...       ...                                                ...\n",
       "27878  K28619                 MPP6; MAGUK p55 subfamily member 6\n",
       "27879  K28620                 MPP7; MAGUK p55 subfamily member 7\n",
       "27880  K28621             CABLES; CDK5 and ABL1 enzyme substrate\n",
       "27881  K28622  LRRTM4; leucine-rich repeat transmembrane neur...\n",
       "27882  K28623             MRNIP; MRN complex-interacting protein\n",
       "\n",
       "[27883 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Consultando la base de ortologos\n",
    "result = REST.kegg_list(database=\"orthology\").read()\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None, sep=\"\\t\")\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4231b938",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>map00010</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>map00020</td>\n",
       "      <td>Citrate cycle (TCA cycle)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>map00030</td>\n",
       "      <td>Pentose phosphate pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>map00040</td>\n",
       "      <td>Pentose and glucuronate interconversions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>map00051</td>\n",
       "      <td>Fructose and mannose metabolism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>map07231</td>\n",
       "      <td>Sodium channel blocking drugs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>map07232</td>\n",
       "      <td>Potassium channel blocking and opening drugs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>map07233</td>\n",
       "      <td>Ion transporter inhibitors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>map07234</td>\n",
       "      <td>Neurotransmitter transporter inhibitors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>map07235</td>\n",
       "      <td>N-Methyl-D-aspartic acid (NMDA) receptor antag...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>581 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            0                                                  1\n",
       "13   map00010                       Glycolysis / Gluconeogenesis\n",
       "14   map00020                          Citrate cycle (TCA cycle)\n",
       "15   map00030                          Pentose phosphate pathway\n",
       "16   map00040           Pentose and glucuronate interconversions\n",
       "17   map00051                    Fructose and mannose metabolism\n",
       "..        ...                                                ...\n",
       "564  map07231                      Sodium channel blocking drugs\n",
       "565  map07232       Potassium channel blocking and opening drugs\n",
       "567  map07233                         Ion transporter inhibitors\n",
       "568  map07234            Neurotransmitter transporter inhibitors\n",
       "566  map07235  N-Methyl-D-aspartic acid (NMDA) receptor antag...\n",
       "\n",
       "[581 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Consultando la base de datos de rutas metabolicas (pathways)\n",
    "\n",
    "result = REST.kegg_list(database=\"pathway\").read()\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df.sort_values(by=0))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c4b7d04",
   "metadata": {},
   "source": [
    "Para **restringir los resultados** únicamente a aquellas rutas metabólicas (*pathways*) presentes en ***drosophila melanogaster***, podemos filtrar los resultados de la base de datos utilizando el código del organismo **dme** como segundo argumento."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d2ece452",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>dsp:122128409</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Slc16a6; monocarboxylate transporter 7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dsp:122094919</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Llgl2; LLGL scribble cell polarity complex com...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dsp:122094920</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Slc16a5; monocarboxylate transporter 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>dsp:122094921</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Ddx5; probable ATP-dependent RNA helicase DDX5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dsp:122094922</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>Itgb4; integrin beta-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24739</th>\n",
       "      <td>dsp:122128285</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24740</th>\n",
       "      <td>dsp:122128286</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24741</th>\n",
       "      <td>dsp:122128318</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24742</th>\n",
       "      <td>dsp:122128319</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24743</th>\n",
       "      <td>dsp:122128323</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>U6 spliceosomal RNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24744 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0      1        2  \\\n",
       "0      dsp:122128409    CDS  Unknown   \n",
       "1      dsp:122094919    CDS  Unknown   \n",
       "2      dsp:122094920    CDS  Unknown   \n",
       "3      dsp:122094921    CDS  Unknown   \n",
       "4      dsp:122094922    CDS  Unknown   \n",
       "...              ...    ...      ...   \n",
       "24739  dsp:122128285  ncRNA  Unknown   \n",
       "24740  dsp:122128286  ncRNA  Unknown   \n",
       "24741  dsp:122128318  ncRNA  Unknown   \n",
       "24742  dsp:122128319  ncRNA  Unknown   \n",
       "24743  dsp:122128323  ncRNA  Unknown   \n",
       "\n",
       "                                                       3  \n",
       "0                 Slc16a6; monocarboxylate transporter 7  \n",
       "1      Llgl2; LLGL scribble cell polarity complex com...  \n",
       "2                 Slc16a5; monocarboxylate transporter 6  \n",
       "3         Ddx5; probable ATP-dependent RNA helicase DDX5  \n",
       "4                                 Itgb4; integrin beta-4  \n",
       "...                                                  ...  \n",
       "24739                                U6 spliceosomal RNA  \n",
       "24740                                U6 spliceosomal RNA  \n",
       "24741                                U6 spliceosomal RNA  \n",
       "24742                                U6 spliceosomal RNA  \n",
       "24743                                U6 spliceosomal RNA  \n",
       "\n",
       "[24744 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Rutas metabolicas (pathways), presentes en Drosophila\n",
    "result = REST.kegg_list(database=\"dsp\").read()\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d337e826",
   "metadata": {},
   "source": [
    "Finalmente, con `kegg_list` podemos realizar consultas más específicas, como obtener información de un gen o de una lista de genes en particular, incluso de diferentes organismos, o de un compuesto específico. Por ejemplo:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9cb67c9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gsu:GSU1176</td>\n",
       "      <td>frdC; succinate dehydrogenase/fumarate reducta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C01290</td>\n",
       "      <td>Lactosylceramide; beta-D-Galactosyl-(1-&gt;4)-bet...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0                                                  1\n",
       "0  gsu:GSU1176  frdC; succinate dehydrogenase/fumarate reducta...\n",
       "1       C01290  Lactosylceramide; beta-D-Galactosyl-(1->4)-bet..."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Consultando la información de dos genes, un compuesto y una glicina\n",
    "result = REST.kegg_list([\"gsu:GSU1176\", \"cpd:C01290\"]).read()\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09e41b48",
   "metadata": {},
   "source": [
    "## 3. `KeggInfo`\n",
    "\n",
    "La función `kegg_info()` devuelve información básica sobre una base de datos específica de **KEGG**, de manera similar a visitar la página principal de esa base de datos. Consultemos la ayuda de la función `kegg_ifo`, para conocer la descripción y parámetros."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ef2aba35",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function kegg_info in module Bio.KEGG.REST:\n",
      "\n",
      "kegg_info(database)\n",
      "    KEGG info - Displays the current statistics of a given database.\n",
      "\n",
      "    db - database or organism (string)\n",
      "\n",
      "    The argument db can be a KEGG database name (e.g. 'pathway' or its\n",
      "    official abbreviation, 'path'), or a KEGG organism code or T number\n",
      "    (e.g. 'hsa' or 'T01001' for human).\n",
      "\n",
      "    A valid list of organism codes and their T numbers can be obtained\n",
      "    via kegg_info('organism') or https://rest.kegg.jp/list/organism\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Solicitando la ayuda de kegg_info\n",
    "help(REST.kegg_info)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d78f73a9",
   "metadata": {},
   "source": [
    "### 3.1 Información general de Kegg\n",
    "\n",
    "Para obtener información general de KEGG y sus bases de datos en su conjunto, se puede usar `kegg_info(\"kegg\")`, como se muestra en el siguiente código:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "24e83dd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "kegg             Kyoto Encyclopedia of Genes and Genomes\n",
      "kegg             Release 116.0+/11-04, Nov 25\n",
      "                 Kanehisa Laboratories\n",
      "                 pathway         581 entries\n",
      "                 brite       442,420 entries\n",
      "                 module      675,404 entries\n",
      "                 orthology    27,883 entries\n",
      "                 genome       26,047 entries\n",
      "                 genes     63,589,682 entries\n",
      "                 compound     19,532 entries\n",
      "                 glycan       11,237 entries\n",
      "                 reaction     12,323 entries\n",
      "                 rclass        3,209 entries\n",
      "                 enzyme        8,274 entries\n",
      "                 network       1,683 entries\n",
      "                 variant       1,731 entries\n",
      "                 disease       2,997 entries\n",
      "                 drug         12,714 entries\n",
      "                 dgroup        2,526 entries\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Consultando la información de kegg\n",
    "result = REST.kegg_info(\"kegg\").read()\n",
    "\n",
    "print(result)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f011036f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f2bbacda",
   "metadata": {},
   "source": [
    "La consulta nos devuelve información referente a la base de datos consultada, como la versión, la fecha de liberación, el número de registros, entre otros detalles."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3914db91",
   "metadata": {},
   "source": [
    "### 3.2 Información de una bse de datos específica\n",
    "\n",
    "La función `kegg_info()` es un poco más versátil, ya que también puede obtener información sobre bases de datos específicas.  \n",
    "\n",
    "Por ejemplo, podemos consultar la información relacionada con la base de datos *glycan*, como se muestra a continuación:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "20a609db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "glycan           KEGG Glycan Database\n",
      "gl               Release 116.0+/11-05, Nov 25\n",
      "                 Kanehisa Laboratories\n",
      "                 11,237 entries\n",
      "\n",
      "linked db        pathway\n",
      "                 brite\n",
      "                 module\n",
      "                 compound\n",
      "                 reaction\n",
      "                 enzyme\n",
      "                 pubchem\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Información de la BD glycan\n",
    "result = REST.kegg_info(\"glycan\").read()\n",
    "\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96c441bb",
   "metadata": {},
   "source": [
    "En este caso, además de información como el número de registros, la versión y la fecha de liberación, también obtenemos los nombres de las bases de datos relacionadas con nuestra consulta, así como la **clave de consulta** de la base (gl).\n",
    "\n",
    "Veamos **otro ejemplo**, ahora consultando la información del **organismo Drosophila melanogaster(dme)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8eead133",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "T00030           Drosophila melanogaster (fruit fly) KEGG Genes Database\n",
      "dme              Release 116.0+/10-30, Oct 25\n",
      "                 Kanehisa Laboratories\n",
      "                 15,243 entries\n",
      "\n",
      "linked db        pathway\n",
      "                 brite\n",
      "                 module\n",
      "                 ko\n",
      "                 genome\n",
      "                 enzyme\n",
      "                 ncbi-geneid\n",
      "                 ncbi-proteinid\n",
      "                 uniprot\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Información de la BD relacionada con la información de Drosophila (dme)\n",
    "result = REST.kegg_info(\"dme\").read()\n",
    "print(result)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dff01fd",
   "metadata": {},
   "source": [
    "Como podemos observar, en esta consulta obtenemos información relacionada con el organismo **Drosophila melanogaster**, incluyendo su **clave de consulta** (`dme`), su **identificador de organismo** (**Tnumber: T00030**), entre otros datos relevantes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6f319aa",
   "metadata": {},
   "source": [
    "## 4. `kegg_conv`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90a3005d",
   "metadata": {},
   "source": [
    "`kegg_conv` es una función que permite **convertir identificadores** entre **KEGG** y otras bases de datos externas, como **UniProt**, **NCBI-Gene**, **PubChem**, entre otras.  \n",
    "\n",
    "Esta herramienta resulta muy útil cuando se desea **relacionar información entre diferentes fuentes biológicas**; por ejemplo, para encontrar los identificadores de KEGG correspondientes a una lista de genes de UniProt, o viceversa.  \n",
    "\n",
    "A continuación, consultaremos la ayuda de la función `kegg_conv` para revisar su descripción y los parámetros que utiliza.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e9aee1e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function kegg_conv in module Bio.KEGG.REST:\n",
      "\n",
      "kegg_conv(target_db, source_db, option=None)\n",
      "    KEGG conv - convert KEGG identifiers to/from outside identifiers.\n",
      "\n",
      "    Arguments:\n",
      "     - target_db - Target database\n",
      "     - source_db_or_dbentries - source database or database entries\n",
      "     - option - Can be \"turtle\" or \"n-triple\" (string).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Solicitando la ayuda de kegg_conv\n",
    "help(REST.kegg_conv)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18e81233",
   "metadata": {},
   "source": [
    "### 4.1 Conversión de identificadores de keeg a NCBI\n",
    "\n",
    "Podemos convertir los identificadores de genes de KEGG a sus equivalentes en NCBI-Gene, utilizando como argumento el nombre de la base de datos de a consultar (\"ncbi-geneid\") y el código del organismo de interés. \n",
    "\n",
    "En el siguiente ejemplo, realizamos la conversión para Drosophila melanogaster (dme):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "14c3c169",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Convirtiendo los identificadores de Kegg a sus equivalentes en NCBI\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e5e2839",
   "metadata": {},
   "source": [
    "El resultado es una tabla que muestra las correspondencias entre los **identificadores de KEGG** y sus **identificadores equivalentes en NCBI-Gene** para el organismo dme(Drosophila melanogaster).  \n",
    "\n",
    "> **Nota:** Es importante mencionar que, en este caso en particular, sabemos, por la consulta que realizamos con `REST.kegg_info(\"dme\")`, que el organismo `dme` está vinculado con la base de datos `ncbi-geneid`.  \n",
    "> Sin embargo, si no existiera dicha vinculación, la consulta generaría un **error**.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "913e46ec",
   "metadata": {},
   "source": [
    "Las consultas no necesariamente deben realizarse sobre **todos los genes de un organismo**; también es posible hacerlas sobre **un conjunto específico de genes**, como se muestra a continuación:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "198ae007",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Convirtiendo algunos identificadores de genes a su equivalente de uniprot\n",
    "\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "987ba67a",
   "metadata": {},
   "source": [
    "## 5. `kegg_find`\n",
    "\n",
    "`kegg_find` realiza una **búsqueda de un término dentro de una base de datos específica KEGG**. Esta función permite localizar genes, compuestos, rutas metabólicas, enfermedades u otros elementos que coincidan con el término de búsqueda.\n",
    "\n",
    "Consultemos la ayuda de la función para verificar su **descripción** y los **parámetros disponibles**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "529d13c5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function kegg_find in module Bio.KEGG.REST:\n",
      "\n",
      "kegg_find(database, query, option=None)\n",
      "    KEGG find - Data search.\n",
      "\n",
      "    Finds entries with matching query keywords or other query data in\n",
      "    a given database.\n",
      "\n",
      "    db - database or organism (string)\n",
      "    query - search terms (string)\n",
      "    option - search option (string), see below.\n",
      "\n",
      "    For the compound and drug database, set option to the string 'formula',\n",
      "    'exact_mass' or 'mol_weight' to search on that field only. The\n",
      "    chemical formula search is a partial match irrespective of the order\n",
      "    of atoms given. The exact mass (or molecular weight) is checked by\n",
      "    rounding off to the same decimal place as the query data. A range of\n",
      "    values may also be specified with the minus(-) sign.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Solicitando la ayuda de kegg_find\n",
    "help(REST.kegg_find)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e047a6ff",
   "metadata": {},
   "source": [
    "### 5.1 Búsqueda de un término en particular\n",
    "\n",
    "Podemos buscar si un término específico, como `\"cytochrome\"`, se encuentra en la base de datos de un organismo en particular, por ejemplo *Drosophila melanogaster* (`dme`).\n",
    "Además, podemos realizar una búsqueda más general en la base de datos de pathway o gene. El siguiente código muestra ambos casos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fd1265f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Busqueda en organismo dme ----\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Busqueda en la BD de pathway ----\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Buscando un termino particular (cytochrome) en un organismo (dme)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "print(\"--- Busqueda en organismo dme ----\")\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)\n",
    "\n",
    "# Buscando el mismo término (cytochrome) en la base de datos de pathway\n",
    "\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "print(\"--- Busqueda en la BD de pathway ----\")\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8578af68",
   "metadata": {},
   "source": [
    "### 5.2 Búsqueda de múltiples términos\n",
    "\n",
    "Con la consulta anterior, KEGG devuelve información para la entrada que hemos solicitado. Sin embargo, también podemos usar **múltiples términos de búsqueda** combinándolos con el símbolo +, lo que obliga a que todos los términos especificados estén presentes en los resultados. Por ejemplo, para buscar los términos `toxin` y `venom`, en la **BD** de `genes`,se podría usar la consulta:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "57e36fd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Buscando dos terminos (toxin, venom) en la base de datos de \"genes\"\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b533918",
   "metadata": {},
   "source": [
    "### 5.3 Otros ejemplos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5a152f3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "======== Peso molecular de 300 ===========\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "======== Masa en rango de 174.05-180.07 ===========\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "======= Fórmula en C7H10O5 ============\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Buscando el peso molecular de 300 en la base de datos de compuestos\n",
    "print(\"======== Peso molecular de 300 ===========\")\n",
    "\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)\n",
    "\n",
    "# Buscando la masa en un rango de 174.05-180.07 en la base de datos de compuestos\n",
    "print(\"======== Masa en rango de 174.05-180.07 ===========\")\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)\n",
    "\n",
    "# Buscando la fórmula en C7H10O5 en la base de datos de compuestos\n",
    "print(\"======= Fórmula en C7H10O5 ============\")\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fe74cc6",
   "metadata": {},
   "source": [
    "## 6. `kegg_get`\n",
    "\n",
    "`kegg_get` permite **recuperar toda la información asociada a un identificador de KEGG** (por ejemplo, un gen, compuesto o ruta metabólica).\n",
    "Solicitaremos la ayuda de la función para conocer su **descripción, parámetros** y los **valores que puede devolver.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9660f246",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function kegg_get in module Bio.KEGG.REST:\n",
      "\n",
      "kegg_get(dbentries, option=None)\n",
      "    KEGG get - Data retrieval.\n",
      "\n",
      "    dbentries - Identifiers (single string, or list of strings), see below.\n",
      "    option - One of \"aaseq\", \"ntseq\", \"mol\", \"kcf\", \"image\", \"kgml\" (string)\n",
      "\n",
      "    The input is limited up to 10 entries.\n",
      "    The input is limited to one pathway entry with the image or kgml option.\n",
      "    The input is limited to one compound/glycan/drug entry with the image option.\n",
      "\n",
      "    Returns a handle.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Solicitando la ayuda de kegg_get\n",
    "help(REST.kegg_get)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b0d9e97",
   "metadata": {},
   "source": [
    "Por ejemplo, podemos obtener la información del gen `dme:Dmel_CG13865`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f5dd38fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Obteniendo la información del gen dme:Dmel_CG13865\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f43b43b7",
   "metadata": {},
   "source": [
    "Como podemos observar, el resultado de la consulta es un **texto plano**, que no sigue el formato de una tabla con el mismo número de renglones y columnas, por lo tanto no lo convertiremos a dataframe.\n",
    "\n",
    "Para acceder a cada elemento (por ejemplo, `ENTRY`, `SYMBOL`, `NAME`, etc.), debemos **parsear** el archivo y extraer la información que necesitamos, dependiendo también del tipo de salida que obtenemos.  \n",
    "\n",
    "Por ejemplo, si solicitamos la información del compuesto `cpd:C00051`, obtendremos la siguiente información:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d3335583",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consultando la información del compuesto cpd:C00051\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c213480e",
   "metadata": {},
   "source": [
    "En este caso, además de las etiquetas `ENTRY` y `NAME`, obtenemos etiquetas como `FORMULA`, `EXACT_MASS`, `MOL_WEIGHT`, entre otras.  \n",
    "\n",
    "El resultado se presenta también como un **texto plano**, que debe ser **parseado** para extraer la información deseada.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e63f1249",
   "metadata": {},
   "source": [
    "### 6.1 Recuperación de datos específicos (secuencias, estructuras, imágenes) con `kegg_get`\n",
    "\n",
    "La función `kegg_get` permite, además de proporcionar un identificador de **KEGG**, incluir un **valor opcional** (`\"aaseq\"`, `\"ntseq\"`, `\"mol\"`, `\"kcf\"`, `\"image\"`, `\"kgml\"`) para consultar información específica, como la **secuencia de un gen**, la **imagen de una ruta metabólica** (*pathway*), entre otras.  \n",
    "\n",
    "Por ejemplo, para consultar la **estructura molecular** del compuesto `cpd:C00051`, utilizamos la opción `\"image\"`. De esta manera, obtenemos únicamente la imagen correspondiente a la estructura del compuesto.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6f73c6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cargando el modulo Image de la biblioteca IPython.display, para desplegar imágenes\n",
    "from IPython.display import Image\n",
    "\n",
    "# Consultando la estructura molecular del compuesto cpd:C00051\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ed87a33",
   "metadata": {},
   "source": [
    "Otro ejemplo, es **obtener** solo la **secuencia de aminoácidos** del gen **\"dme:Dmel_CG13865\"**, utilizando la opción \"aaseq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "e8312c4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consultando la secuencia de aminoácidos del gen Dmel_CG13865\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1af0e974",
   "metadata": {},
   "source": [
    "### 6.2 Recuperando imágenes de rutas metabólicas en KEGG\n",
    "\n",
    "`KEGG` es ampliamente reconocido por sus diagramas de rutas metabólicas (*pathways*), los cuales representan de forma visual las interacciones y procesos biológicos.\n",
    "\n",
    "Podemos **obtener estas imágenes directamente desde Python** utilizando la función `kegg_get()` con la opción `\"image\"`. Esta opción permite descargar la representación gráfica de una ruta metabólica específica.  \n",
    "\n",
    "Para especificar un mapa de una ruta en particular, se combina el prefijo `map` con el número de la ruta metabólica que se desea consultar. Por ejemplo, el mapa correspondiente a **fatty-acid biosynthesis** sería `map00061`, el cual se puede solicitar como imagen, como se muestra a continuación:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8a8a29ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consultando el mapa correspondiente a la ruta fatty-acid biosynthesis map00061\n",
    "\n",
    "\n",
    "\n",
    "# Otro ejemplo es el mapa del metabolismo central \"map01100\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d6961e8",
   "metadata": {},
   "source": [
    "Si queremos recuperar el mapa de una ruta metabólica **correspondiente a un organismo en particular**, podemos reemplazar el prefijo `map` por el **código de tres letras** asignado a ese organismo.  \n",
    "\n",
    "Por ejemplo, en el caso de *Drosophila melanogaster*, el prefijo `map` se reemplaza por `dme`, como se muestra en los siguientes ejemplos:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e964a629",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consultando el mapa correspondiente a la ruta fatty-acid biosynthesis map00061 del organismo dme(Drosophila)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf3ed1df",
   "metadata": {},
   "source": [
    "## 7. `kegg_link`\n",
    "\n",
    "La función `kegg_link` se utiliza para **encontrar relaciones o enlases entre diferentes bases de datos de Kegg**. Solicitaremos la ayuda de esta función para conocer la **descripción, parámetros y sintaxis** de la función, antes de revisar algunos ejemplos prácticos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9d9801d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function kegg_link in module Bio.KEGG.REST:\n",
      "\n",
      "kegg_link(target_db, source_db, option=None)\n",
      "    KEGG link - find related entries by using database cross-references.\n",
      "\n",
      "    target_db - Target database\n",
      "    source_db_or_dbentries - source database\n",
      "    option - Can be \"turtle\" or \"n-triple\" (string).\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Solicitando la ayuda de kegg_link\n",
    "help(REST.kegg_link)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9455ba7e",
   "metadata": {},
   "source": [
    "Podemos **relacionar la información de las rutas metabólicas y los genes de un organismo**, con el propósito de identificar en qué rutas participa cada gen. Por ejempo, consultemos la relación entre los genes de *Drosophila melanogaster* y las rutas metabólicas(*pathway*)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "81f058c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T00030           Drosophila melanogaster (frui...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>dme              Release 116.0+/10-30, Oct 25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kanehisa Laboratories</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15,243 entries</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>linked db        pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>brite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>module</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ko</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ncbi-geneid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ncbi-proteinid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>uniprot</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   T00030           Drosophila melanogaster (frui...\n",
       "1       dme              Release 116.0+/10-30, Oct 25\n",
       "2                               Kanehisa Laboratories\n",
       "3                                      15,243 entries\n",
       "4                            linked db        pathway\n",
       "5                                               brite\n",
       "6                                              module\n",
       "7                                                  ko\n",
       "8                                              genome\n",
       "9                                              enzyme\n",
       "10                                        ncbi-geneid\n",
       "11                                     ncbi-proteinid\n",
       "12                                            uniprot"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Consultando la relación de Drosophila con las rutas metabólicas\n",
    "\n",
    "\n",
    "\n",
    "# Transformando los resultados a un dataframe\n",
    "df = pd.read_table(io.StringIO(result), header=None)\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab6c3106",
   "metadata": {},
   "source": [
    "## 8. Puntos clave\n",
    "\n",
    "En esta clase aprendimos a **consultar información en la base de datos de KEGG** utilizando **Biopython**, en particular el módulo `Bio.KEGG.REST`.  \n",
    "\n",
    "Las funciones que utilizamos:\n",
    "\n",
    "- **`kegg_list`** → Muestra los elementos disponibles en una base de datos de KEGG, como organismos, compuestos o rutas metabólicas.  \n",
    "- **`kegg_info`** → Proporciona información general o estadística sobre una base de datos o sobre KEGG en general.  \n",
    "- **`kegg_conv`** → Convierte identificadores entre KEGG y otras bases de datos (por ejemplo, UniProt o NCBI).  \n",
    "- **`kegg_find`** → Permite buscar términos o palabras clave dentro de una base de datos.  \n",
    "- **`kegg_get`** → Recupera la información completa asociada a un identificador (como un gen, compuesto o ruta).  \n",
    "- **`kegg_link`** → Muestra las relaciones entre diferentes bases de datos (por ejemplo, entre genes y rutas metabólicas).\n",
    "\n",
    "En resumen, aprendimos a **explorar y relacionar la información biológica en KEGG** desde Python, entendiendo cómo están organizadas las bases de datos y cómo acceder a la información que necesitamos de manera práctica.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf702c62",
   "metadata": {},
   "source": [
    "## 9. Ejercicios propuestos\n",
    "\n",
    "1. **Crear una función para parsear los resultados de `kegg_get`.**  \n",
    "   Desarrollar una función que permita *parsear* los resultados obtenidos con `kegg_get`, extrayendo todos los campos clave de la consulta (`ENTRY`, `NAME`, entre otros) y la información asociada a cada uno de ellos, **independientemente del tipo de identificador** consultado (gen, compuesto, ruta metabólica, etc.).\n",
    "\n",
    "2. **Recuperar información específica de un campo clave.**  \n",
    "   Implementar una función que permita recuperar la información de un campo clave en particular (por ejemplo, `BRITE`).  \n",
    "   Considere que algunos campos pueden obtenerse directamente agregando un parámetro opcional a la función `kegg_get` —como `\"aaseq\"`, `\"ntseq\"`, `\"mol\"`, `\"kcf\"`, `\"image\"` o `\"kgml\"`.  \n",
    "   Sin embargo, para los campos que **no** se encuentren entre estos parámetros opcionales, deberá utilizar la función del **Ejercicio 1** para parsear el resultado y obtener la información deseada.\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".main_mamba",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.14.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
